mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/BlockMatrix.scala

```diff
@@ -268,8 +268,26 @@ class BlockMatrix @Since("1.3.0") (
   def toIndexedRowMatrix(): IndexedRowMatrix = {
     require(numCols() < Int.MaxValue, "The number of columns must be within the integer range. " +
       s"numCols: ${numCols()}")
-    // TODO: This implementation may be optimized
-    toCoordinateMatrix().toIndexedRowMatrix()
+
+    val rows = blocks.map(block => (block._1._1, (block._1._2, block._2)))
+      .groupByKey()
```

Comment:
Thanks @mengxr, that's a very good idea. I'll update the code and push it within 24 hours. Cheers!